// bitcoin-ts/build/main/lib/crypto/hash.browser.bench.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:no-expression-statement no-unsafe-any
const ava_1 = __importDefault(require("ava"));
const path_1 = require("path");
const puppeteer_1 = require("puppeteer");
const rollup_1 = require("rollup");
const rollup_plugin_alias_1 = __importDefault(require("rollup-plugin-alias"));
const rollup_plugin_commonjs_1 = __importDefault(require("rollup-plugin-commonjs"));
const rollup_plugin_node_resolve_1 = __importDefault(require("rollup-plugin-node-resolve"));
const prepareCode = async () => {
    // tslint:disable-next-line:no-unbound-method no-console
    const realConsoleWarn = console.warn;
    /**
     * Suppress Rollup warning: `Use of eval is strongly discouraged, as it poses
     * security risks and may cause issues with minification`
     */
    // tslint:disable-next-line:no-object-mutation no-console
    console.warn = (suppress) => suppress;
    const bundle = await rollup_1.rollup({
        // TODO: remove after https://github.com/rollup/rollup/pull/2348 lands
        inlineDynamicImports: false,
        input: path_1.join(__dirname, 'hash.browser.bench.helper.js'),
        plugins: [
            rollup_plugin_alias_1.default({
                chuhai: './../../../bench/chuhai.js',
                'hash.js': './../../../bench/hash.js'
            }),
            rollup_plugin_commonjs_1.default(),
            rollup_plugin_node_resolve_1.default()
        ]
    });
    // tslint:disable-next-line:no-object-mutation no-console
    console.warn = realConsoleWarn;
    const { code } = await bundle.generate({
        format: 'esm'
    });
    return code;
};
const preparePage = async () => {
    const browser = await puppeteer_1.launch({
        args: ['--no-sandbox', '--disable-setuid-sandbox']
        // devtools: true
    });
    const page = await browser.newPage();
    // https://github.com/GoogleChrome/puppeteer/issues/2301#issuecomment-379622459
    await page.goto('file:///');
    return { browser, page };
};
(async () => {
    const [code, { browser, page }] = await Promise.all([
        prepareCode(),
        preparePage()
    ]);
    ava_1.default(`# browser: ${await browser.version()}`, async (t) => {
        page.on('console', msg => {
            // tslint:disable-next-line:no-console
            console.log(msg.text());
        });
        page.on('error', err => {
            // tslint:disable-next-line:no-console
            console.error(`error: ${err}`);
        });
        page.on('pageerror', err => {
            // tslint:disable-next-line:no-console
            console.error(`pageerror: ${err}`);
        });
        await new Promise(async (resolve) => {
            await page.exposeFunction('benchError', (error) => {
                // tslint:disable-next-line:no-console
                console.error(error);
            });
            await page.exposeFunction('benchComplete', async () => {
                // tslint:disable-next-line:no-console
                console.log('Browser benchmark complete, closing browser.');
                await browser.close();
                t.pass();
                resolve();
            });
            await page.setContent(`<script type="module">${code}</script>`);
        });
    });
})().catch(err => {
    // tslint:disable-next-line:no-console
    console.error(err);
});
