// bitcoin-ts/build/main/lib/crypto/sha256.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bin_1 = require("../bin/bin");
/**
 * The most performant way to instantiate sha256 functionality. To avoid
 * using Node.js or DOM-specific APIs, you can use `instantiateSha256`.
 *
 * @param webassemblyBytes A buffer containing the sha256 binary.
 */
exports.instantiateSha256Bytes = async (webassemblyBytes) => {
    const wasm = await bin_1.instantiateRustWasm(webassemblyBytes, './sha256', 'sha256', 'sha256_init', 'sha256_update', 'sha256_final');
    return {
        final: wasm.final,
        hash: wasm.hash,
        init: wasm.init,
        update: wasm.update
    };
};
exports.getEmbeddedSha256Binary = () => bin_1.decodeBase64String(bin_1.sha256Base64Bytes);
/**
 * An ultimately-portable (but slower) version of `instantiateSha256Bytes`
 * which does not require the consumer to provide the sha256 binary buffer.
 */
exports.instantiateSha256 = async () => exports.instantiateSha256Bytes(exports.getEmbeddedSha256Binary());
